compiler/luci/pass/src/QuantizeDequantizeWeightsPass.cpp

```diff
@@ -70,6 +70,198 @@ void compute_asym_scale_zp(float min, float max, float &scaling_factor, int64_t
+      len_per_dim[2] = node->dim(2).value();
+      len_per_dim[3] = node->dim(3).value();
+      channel_dim_index = 3;
+      return true;
+    }
+    else if ((tw_conv != nullptr && tw_conv->filter() == node) ||
+             (dw_conv != nullptr && dw_conv->filter() == node))
+    {
+      len_per_dim[0] = node->dim(0).value();
+      len_per_dim[1] = node->dim(1).value();
+      len_per_dim[2] = node->dim(2).value();
+      len_per_dim[3] = node->dim(3).value();
+      channel_dim_index = 2;
+      return true;
+    }
+    else if (fc != nullptr && fc->weights() == node)
+    {
+      len_per_dim[0] = 1;
+      len_per_dim[1] = 1;
+      len_per_dim[2] = node->dim(0).value();
+      len_per_dim[3] = node->dim(1).value();
+      channel_dim_index = 3;
+      return true;
+    }
+    else
+    {
+      // node does not support channle-wise quantization
+      asset(false);
+    }
+  }
+
+  return false;
+}
+
+int cal_offset(int *len_per_dim, int *indices)
+{
+  return indices[0] * len_per_dim[1] * len_per_dim[2] * len_per_dim[3] +
+         indices[1] * len_per_dim[2] * len_per_dim[3] + indices[2] * len_per_dim[3] + indices[3];
+}
+
+void cal_minmax_per_channel(CircleConst *node, std::vector<float> &min, std::vector<float> &max)
+{
+  const int len = 4;
+  static_assert(len == 4, "all weights should be set to the same rank as the convolution weights");
+  int len_per_dim[len];
+  int indices[len];
+  int channel_dim_index;
+  int size;
+
+  bool ret = get_channel_dim_index(node, len_per_dim, channel_dim_index);
+  assert(ret);
+  size = len_per_dim[channel_dim_index];
+
+  std::vector<bool> has_min_max_value(size, false);
+  min.resize(size);
+  max.resize(size);
+  for (indices[0] = 0; indices[0] < len_per_dim[0]; indices[0]++)
+  {
+    for (indices[1] = 0; indices[1] < len_per_dim[1]; indices[1]++)
+    {
+      for (indices[2] = 0; indices[2] < len_per_dim[2]; indices[2]++)
+      {
+        for (indices[3] = 0; indices[3] < len_per_dim[3]; indices[3]++)
+        {
+          int channel_idx = indices[channel_dim_index];
+          auto data = node->at<loco::DataType::FLOAT32>(cal_offset(len_per_dim, indices));
+          if (has_min_max_value[channel_idx])
+          {
+            min[channel_idx] = data < min[channel_idx] ? data : min[channel_idx];
+            max[channel_idx] = data > max[channel_idx] ? data : max[channel_idx];
+          }
+          else
+          {
+            min[channel_idx] = data;
+            max[channel_idx] = data;
+            has_min_max_value[channel_idx] = true;
+          }
+        }
+      }
+    }
+  }
+}
+
+void wquant_per_channel(CircleConst *node, std::vector<float> &min, std::vector<float> &max,
+                        std::vector<float> &scaling_factor, std::vector<int64_t> &zp,
+                        std::vector<float> &nudged_min, std::vector<float> &nudged_max)
+{
+  const int32_t kMinScale = 0;
+  const int32_t kMaxScale = 255;
+
+  uint32_t size = node->size<loco::DataType::FLOAT32>();
```

Comment:
I think there is a precondition that `assert(node->dtype() ==loco::DataType::FLOAT32);`
can you please add this `assert` at the top of this function?

I think `wdequant_per_channel` will need this too.

Reply:
I applied it line number ```184```